.meta/sources/docker.toml

```diff
@@ -55,6 +55,24 @@ A list of image names to match against. If not provided, \
 all images will be included.\
 """
 
+[sources.docker.options.no_auto_partial_merge]
+type = "bool"
+common = false
+default = false
+description = """\
+Setting this to `false` will disable the automatic merging of partial events.\
+"""
+
+[sources.docker.options.partial_event_marker]
```

Comment:
I was just thinking about the high-level UX for the `docker` source, and I'm wondering why we even need to expose this option to the user? The `no_auto_partial_merge` option handles merging so they'll never be exposed to the `partial_event_marker` field. The intracies involved in merging are completely hidden from the user. Alternatively, if the user disables automatic merging I can't think of a use case where the user would want to continue adding the `partial_event_marker` field. If they plan to pair this with the `merge` transform they might as well enable automatic merging. Does that make sense?

Reply:
Actually, this is a very good point. The reason I designed it like this is to provide an opportunity for the user to add additional logic to alter the partial messages before they are merged together.

It is hard for me to find a meaningful way of altering partial events in a meaningful way though. This might actually be a general concern with the feature, not just with the `docker` source case. The design was built around providing a way to access partial messages before doing the merge. One example where this can actually be useful to do a (semi-)meaningful thing is unmarking some messages in a partial message chain on certain conditions, the other is dropping partial events. Both can be implemented in a composable way, outside the merge transform itself. The problem is they're both quite simple, are related exclusively to the message flow management and have a very simple outcome - and that makes me think we could just implement them as part of the merge transform itself.

I guess it'd be a good idea to brainstorm a bit about the possible applications of the partial message marking and merging being separate units in the topology, compared to them both being contained in the merge transform (i.e. it would do both partial message marking and merging). We can maintain the flexibility of the partial events marking configuration we provide currently by accepting arbitrary transforms as arguments to the merge transform to allow users to implement arbitrarily complex logic (yet, less complex than would be possible with a separate marking transforms).

Getting back to the `docker` source - I agree with what you wrote, this makes a ton of sense. We have a benefit of knowing the data we're working with here, and it for sure looks like the use cases for adding the partial event marker here is void cause it's otherwise covered by the automatic merging.

It is possible to get rid of the partial marker field entirely by reimplementing the partial marking and merging logic in a custom way that doesn't directly use the merge transform - but just reuses the event merging utilities it provides. This would eliminate the headache of dealing with field collisions - it's a minor issue, but getting rid of it just nicely reduces the cognitive load for the users.

Reply:
> It is possible to get rid of the partial marker field entirely by reimplementing the partial marking and merging logic in a custom way that doesn't directly use the merge transform - but just reuses the event merging utilities it provides. This would eliminate the headache of dealing with field collisions - it's a minor issue, but getting rid of it just nicely reduces the cognitive load for the users.

Exactly. I think that's what we're leaning towards. The merge abstraction will be more clear once we have a few sources supporting merging (`docker`, `file`, `socket`, and `journald`). At that point, we should understand better how we can share code to implement this across all sources.

What do you think? Again, I don't think this is wasted effort, but perhaps we should solve `docker` _only_ first, then proceed with a couple of more sources, and then revisit this PR. I'm willing to bet we'll reuse code here or take inspiration from the work done here.

Reply:
I've addressed this at https://github.com/timberio/vector/pull/1504#issuecomment-577015837

I'd like to have `merge` transform too, but the only concern I have is that it'd be an unsettled, `unstable` interface if we merge it now.
That said, I'd prefer to have it merged rather than for it to be stashed for later, cause it really helps when doing the research on in-source merging mechanisms. You can quickly outline the projected in-source implementation by using a `merge` transform to test-drive it before implementing. This saves some time.

Reply:
@MOZGIII we still need to remove this option, correct?

Reply:
It's still going to be used if the automatic partial merge is disabled. We should leave it, but it's been renamed to `partial_event_marker_field`.